ikarus/finiteelements/mechanics/enhancedassumedstrains.hh

```diff
@@ -138,144 +153,241 @@ public:
     easApplicabilityCheck();
-    using namespace Dune;
-    using namespace Dune::DerivativeDirections;
-    const auto uFunction      = underlying().displacementFunction(par, dx);
-    auto strainFunction       = underlying().strainFunction(par, dx);
-    const auto& numberOfNodes = underlying().numberOfNodes();
+    auto strainFunction     = underlying().strainFunction(par, dx);
+    const auto fIntFunction = underlying().template internalForcesFunction<ScalarType>(par, force, dx);
 
     auto calculateForceContribution = [&]<typename EAST>(const EAST& easFunction) {
-      typename EAST::DType D;
-      calculateDAndLMatrix(easFunction, par, D, L_);
-
-      decltype(auto) LMat = [this]() -> decltype(auto) {
-        if constexpr (Concepts::AutodiffScalar<ScalarType>)
-          return Eigen::MatrixX<ScalarType>{};
-        else
-          return [this]() -> Eigen::MatrixXd& { return L_; }();
-      }();
-      const auto disp  = Dune::viewAsFlatEigenVector(uFunction.coefficientsRef());
-      const auto alpha = (-D.inverse() * L_ * disp).eval();
-      const auto geo   = underlying().localView().element().geometry();
-      auto C           = underlying().materialTangentFunction(par);
-
       for (const auto& [gpIndex, gp] : strainFunction.viewOverIntegrationPoints()) {
-        const auto M            = easFunction.calcM(gp.position());
-        const double intElement = geo.integrationElement(gp.position()) * gp.weight();
-        const auto CEval        = C(gpIndex);
-        auto stresses           = (CEval * M * alpha).eval();
-        for (size_t i = 0; i < numberOfNodes; ++i) {
-          const auto bopI = strainFunction.evaluateDerivative(gpIndex, wrt(coeff(i)), on(gridElement));
-          force.template segment<Traits::worlddim>(Traits::worlddim * i) += bopI.transpose() * stresses * intElement;
-        }
+        const auto enhancedStrain = enhancedStrainFunction(easFunction, strainFunction, gp.position());
+        fIntFunction(enhancedStrain, gpIndex, gp);
+      }
+      if constexpr (EAST::enhancedStrainSize != 0) { // compile-time check
+        typename EAST::DType D;
+        calculateDAndLMatrix(easFunction, par, D, L_);
+        const auto& Rtilde = calculateRtilde(par, dx);
+        force -= L_.transpose() * D.inverse() * Rtilde;
       }
     };
     easVariant_(calculateForceContribution);
   }
 
+  template <typename MT, typename BC>
+  void subscribeToImpl(BC& bc) {
+    if constexpr (std::same_as<MT, NonLinearSolverMessages>) {
+      using NLSState = typename BC::State;
+      underlying().subscribe(bc, [&](NonLinearSolverMessages message, const NLSState& state) {
+        if (message == NonLinearSolverMessages::CORRECTION_UPDATED)
+          this->updateStateImpl(state.domain, state.correction);
+      });
+    }
+  }
+
 private:
-  EAS::Impl::EASVariant<Geometry> easVariant_;
+  EAS::EASVariant<ES, Geometry> easVariant_;
   mutable Eigen::MatrixXd L_;
+  mutable Eigen::VectorXd alpha_;
```

Comment:
This mutable can be removd

Reply:
[Here](https://github.com/ikarus-project/ikarus/blob/1aeece7a99985f2bfbcdb837b3996960ccdd9248/ikarus/finiteelements/mechanics/enhancedassumedstrains.hh#L136) in `calculateAt` function which is marked with `const`, `alpha_` is updated and hence `mutable` is used. We did not yet explore on how to make use of `updateStateImpl(...)` to update `alpha_` correcly to calculate stress here and hence will do it separately later as mentioned [here](https://github.com/ikarus-project/ikarus/pull/325#issue-2535743230). Is this fine?

Reply:
It is somehow a violation of the const correctness that `calculateAt` modifies, the internal alpha. It also does this with a `=` sign sicne the full displacements are given as arguments and in `updateStateImpl` it is a `-=`. Thus, Im wondering if we need that updateState at all in that case and why are we not using it as in `calculateAt`, where the alphas are calculated on the fly? I suppose one of these usages might be wrong? This might be and issue from which iteration the alhpas or dispalcements come from? Can you elaborate @henrij22 @tarun-mitruka ?

Reply:
For now, we can also define a new variable `alpha` in `calculateAtImpl` used only for post-processing and then remove the `mutable` for the member variable `alpha_`. What do you think?

Reply:
I added a new variable in `calculateAtImpl` and removed the `mutable`. Deeper investigation on `calculateAtImpl` will be done in a separate PR. Resolving this now